src/types.ts

```diff
@@ -1,8 +1,79 @@
 import * as mongoose from 'mongoose';
 
-import type { Base } from './defaultClasses';
+import type { Base, TimeStamps } from './defaultClasses';
 import type { Severity, WhatIsIt } from './internal/constants';
 
+// tslint:disable:ban-types
+
+type IfEquals<X, Y, A, B> =
+  (<T>() => T extends X ? 1 : 2) extends
+  (<T>() => T extends Y ? 1 : 2) ? A : B;
+
+type ReadonlyKeysOf<T> = {
+  [P in keyof T]: IfEquals<{ [Q in P]: T[P] }, { -readonly [Q in P]: T[P] }, never, P>
+}[keyof T];
+
+type FunctionPropertyNames<T> = { [K in keyof T]: T[K] extends Function ? K : never }[keyof T];
+
+export interface TypegooseModel<
+  T extends mongoose.Document,
+  QueryHelpers = {},
+  V = Omit<T,
+    ReadonlyKeysOf<T>
+    | FunctionPropertyNames<T>
+    | keyof mongoose.Document
+    | keyof IObjectWithTypegooseFunction
+    | (T extends TimeStamps ? keyof TimeStamps : never)> & Partial<Pick<mongoose.Document, '_id' | '__v'>>,
+  D = { [K in keyof V]: V[K] extends Map<infer X, infer Y> ?
```

Comment:
could you explain to me what this generic does?

Reply:
The param V takes the document type (i.e. the class from `getModelFromClass()` combined with `mongoose.Document`) and removes keys that aren't required to specify. That is, the `Readonly` keys (I did this for set-only properties - consider removing the `ReadonlyKeysOf<T>` part), the functions keys, the keys that are from `mongoose.Document` itself (like `isNew`, `increment`, etc.), the `typegooseName`, and the timestamp data (if the schema extends the `TimeStamps` class). It also allows specifying the `_id` and `__v` from `mongoose.Document`.

The `D` type allows support for map props. It's a bit flimsy, but it takes all of the `Map<X, Y>` properties in the original, assumes they are `mapProp`s, and allows you to, instead of specifying a `Map` object in the constructor, specify an object literal (i.e. a `Record`).

Let me know if you need more clarification.

Reply:
arnt mongoose maps defined as `{ mapprop: [['key1', value], ['key2', value] ] }`?

Reply:
ok nvm, didnt know it can be created via an object too

Reply:
could you change the types to allow the approach i listed, because its supported by mongoose?
(added an test in 46b96ee966798be44490b511bb5c2d13ee9bd5e3)

Reply:
Just pushed in 1f440b26a2cfa12cff8c4a2b438bd363d1c14366.